File: src/mcp/endpoints/agent.py
"""
Agent REST Endpoints

Agent-related HTTP endpoints for Desktop frontend.
"""

import logging

from starlette.requests import Request
from starlette.responses import JSONResponse

from src.mcp.contract import _guard_detection_evasion_payload

logger = logging.getLogger("niko-gateway")


async def agent_route_endpoint(request: Request):
    """Agent route REST endpoint."""
    from src.mcp.gateway import agent_route

    body = await request.json()
    result = await agent_route(task=body.get("task", ""))
    return JSONResponse(result)


async def agent_write_endpoint(request: Request):
    """Agent write REST endpoint."""
    from src.mcp.gateway import agent_write

    body = await request.json()
    quality_goals_payload = body.get("quality_goals") or body.get("qualityGoals")
    write_kwargs = {
        "scene_card": body.get("scene_card") or {},
        "skills": body.get("skills"),
        "word_target": body.get("word_target", 2000),
        "allow_llm_fallback": body.get("allow_llm_fallback", True),
    }
    if quality_goals_payload is not None:
        write_kwargs["quality_goals"] = quality_goals_payload
    result = await agent_write(**write_kwargs)
    return JSONResponse(result)


async def agent_revise_endpoint(request: Request):
    """Agent revise REST endpoint."""
    from src.mcp.gateway import agent_revise, _guard_detection_evasion_payload

    try:
        body = await request.json()
    except Exception:
        return JSONResponse({"error": "Invalid JSON body"}, status_code=400)

    if not isinstance(body, dict):
        return JSONResponse({"error": "Request body must be an object"}, status_code=400)

    blocked = _guard_detection_evasion_payload(body)
    if blocked is not None:
        return blocked

    feedback_value = body.get("feedback")
    if feedback_value is None:
        feedback_value = {}
    elif not isinstance(feedback_value, dict):
        return JSONResponse({"error": "feedback must be an object"}, status_code=400)

    allow_llm_fallback = bool(body.get("allow_llm_fallback", body.get("allowLlmFallback", True)))

    revise_kwargs = {
        "draft": body.get("draft", ""),
        "feedback": feedback_value,
        "allow_llm_fallback": allow_llm_fallback,
    }
    quality_goals_payload = body.get("quality_goals") or body.get("qualityGoals")
    if quality_goals_payload is not None:
        revise_kwargs["quality_goals"] = quality_goals_payload

    try:
        result = await agent_revise(**revise_kwargs)
        return JSONResponse(result)
    except ValueError as exc:
        return JSONResponse({"error": str(exc)}, status_code=400)
    except RuntimeError as exc:
        message = str(exc)
        status_code = 503 if "LLM" in message else 500
        return JSONResponse({"error": message}, status_code=status_code)
    except Exception as exc:
        logger.error(f"Agent revise endpoint error: {exc}")
        return JSONResponse({"error": str(exc), "status": "error", "endpoint": "/agent/revise"}, status_code=500)


async def agent_context_endpoint(request: Request):
    """Agent context REST endpoint."""
    from src.mcp.gateway import agent_get_context

    body = await request.json()
    result = await agent_get_context(
        scene_info=body.get("scene_info") or {},
        context_types=body.get("context_types"),
    )
    return JSONResponse(result)


__all__ = [
    "agent_route_endpoint",
    "agent_write_endpoint",
    "agent_revise_endpoint",
    "agent_context_endpoint",
]

File: src-ts/mcp/endpoints/agent.ts
/**
 * Agent REST Endpoints
 *
 * Agent-related HTTP endpoints for Desktop frontend.
 * Ported from src/mcp/endpoints/agent.py
 */

import type { HttpRequest, HttpResponse } from '../http-types';
import { jsonResponse, parseBody } from '../http-types';
import { normalizeProjectWorkspaceContext } from '../../project/workspace-model.js';
import { safeResolveWorkspaceRoot } from '../input-validation.js';
import {
  agentRoute,
  agentWrite,
  agentRevise,
  agentGetContext,
} from '../services/agent';

function asRecord(value: unknown): Record<string, unknown> | null {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : null;
}

function readString(value: unknown): string | null {
  if (typeof value !== 'string') return null;
  const trimmed = value.trim();
  return trimmed ? trimmed : null;
}

function resolveWorkspaceRoot(): string {
  return safeResolveWorkspaceRoot();
}

function resolveAgentWriteWorkspace(body: Record<string, unknown>) {
  const workspace = asRecord(body.workspace);
  const identity = asRecord(workspace?.identity);
  const context = asRecord(body.context);
  const workspaceId =
    readString(identity?.workspaceId)
    ?? readString(body.workspace_id)
    ?? readString(body.workspaceId);
  const projectId =
    readString(identity?.projectId)
    ?? readString(body.project_id)
    ?? readString(body.projectId)
    ?? readString(context?.projectId);

  if (!workspace && !workspaceId && !projectId) {
    return undefined;
  }

  return normalizeProjectWorkspaceContext({
    workspace: workspace ?? undefined,
    workspace_id: workspaceId ?? undefined,
    workspaceId: workspaceId ?? undefined,
    project_id: projectId ?? undefined,
    projectId: projectId ?? undefined,
    context: context ?? undefined,
  }, {
    workspaceRoot: resolveWorkspaceRoot(),
  });
}

export async function agentRouteEndpoint(request: HttpRequest): Promise<HttpResponse> {
  const body = parseBody(request) as { task?: string };
  const result = await agentRoute(body.task ?? '');
  return jsonResponse(result);
}

export async function agentWriteEndpoint(request: HttpRequest): Promise<HttpResponse> {
  const body = parseBody(request) as Record<string, unknown>;
  const workspace = resolveAgentWriteWorkspace(body);
  const qualityGoals = (body.quality_goals ?? body.qualityGoals) as
    | Record<string, unknown>
    | undefined;

  const result = await agentWrite({
    scene_card: (body.scene_card as Record<string, unknown>) ?? {},
    skills: body.skills as string[] | undefined,
    word_target: (body.word_target as number) ?? 2000,
    allow_llm_fallback: (body.allow_llm_fallback as boolean) ?? true,
    quality_goals: qualityGoals,
    workspace,
  });

  return jsonResponse(result);
}

export async function agentReviseEndpoint(request: HttpRequest): Promise<HttpResponse> {
  let body: Record<string, unknown>;
  try {
    body = parseBody(request) as Record<string, unknown>;
  } catch {
    return jsonResponse({ error: 'Invalid JSON body' }, 400);
  }

  if (typeof body !== 'object' || body === null) {
    return jsonResponse({ error: 'Request body must be an object' }, 400);
  }

  let feedbackValue = body.feedback;
  if (feedbackValue == null) {
    feedbackValue = {};
  } else if (typeof feedbackValue !== 'object') {
    return jsonResponse({ error: 'feedback must be an object' }, 400);
  }

  const allowLlmFallback = Boolean(
    body.allow_llm_fallback ?? body.allowLlmFallback ?? true
  );

  const qualityGoals = (body.quality_goals ?? body.qualityGoals) as
    | Record<string, unknown>
    | undefined;

  const reviseParams = {
    draft: (body.draft as string) ?? '',
    feedback: feedbackValue as Record<string, unknown>,
    allow_llm_fallback: allowLlmFallback,
    quality_goals: qualityGoals,
  };

  try {
    const result = await agentRevise(reviseParams);
    return jsonResponse(result);
  } catch (exc) {
    const message = exc instanceof Error ? exc.message : String(exc);
    if (message.includes('ValueError')) {
      return jsonResponse({ error: message }, 400);
    }
    const statusCode = message.includes('LLM') ? 503 : 500;
    return jsonResponse({ error: message }, statusCode);
  }
}

export async function agentContextEndpoint(request: HttpRequest): Promise<HttpResponse> {
  const body = parseBody(request) as Record<string, unknown>;
  const result = await agentGetContext(
    (body.scene_info as Record<string, unknown>) ?? {},
    body.context_types as string[] | undefined
  );
  return jsonResponse(result);
}
